### ISSUE TITLE
Allow separate Java classes for GRPC calls of same service

### ISSUE THREAD

**Issue by mdewilde** [yidongnan/grpc-spring-boot-starter#401]
**The problem**

Take the following proto

```
service ProjectService {
  rpc CreateProject (CreateProjectRequest) returns (CreateProjectResponse); 
  rpc UpdateProject (UpdateProjectRequest) returns (UpdateProjectResponse);
}
```
This is compiled to a Java `ProjectServiceBaseImpl`, with a public overridable method for each call.

I'd love to separate these into their own Java classes:

```
@GrpcService
@PreAuthorize("isAuthenticated()")
public class CreateProjectService extends ProjectServiceImplBase {
    @Override
    public void createProject(CreateProjectRequest request, StreamObserver<CreateProjectResponse> responseObserver) {
        // ... 
    }
}

@GrpcService
@PreAuthorize("isAuthenticated()")
public class UpdateProjectService extends ProjectServiceImplBase {
    @Override
    public void updateProject(UpdateProjectRequest request, StreamObserver<UpdateProjectResponse> responseObserver) {
        // ... 
    }
}
```

A project with a similar setup compiles, but it does not actually work: calling the GRPC methods produces unreliable results, with either the method in ProjectServiceImplBase being called, or an overridden version in one of the classes.

**The solution**

It would be nice if:
- overridden GRPC calls could be spread over multiple `@GrpcService` classes
- there was startup checking to flag unsupported configurations (eg. two separate overrides for the same GRPC call)

**Alternatives considered**

Even if it is decided that GRPC calls should not be spread over multiple classes, it would still be a plus if there was additional checking as to unsupported configurations.

**Additional context**

If you are open to the idea, I would be willing to explore the library to see how this could be achieved.

**Comment by ST-DDT** [comment]
> there was startup checking to flag unsupported configurations (eg. two separate overrides for the same GRPC call)

This probably belongs to the same category as #365 (Fixed via #392).

> overridden GRPC calls could be spread over multiple @GrpcService classes

This library/grpc is unable to distinguish which methods are actually implemented and in which bean, so it cannot help you with that.

You can only have a single implementation of a specific `ImplBase`. (grpc-java limitation/feature)

There are three ways around this:

1. Create yet another implementation, add the annotation only there and delegate the impls to the respective impl beans.
2. Instead of implementing the `ImplBase` method you have a special bean per method that you delegate the call to. (Similar to 1)
3. Create your own `BindableService` implementation, that delegates to multiple beans.
4. (Split the service-definition into multiple smaller ones)


-----------

1+2) Would roughly look like this:

````java
@GrpcService
public class MainProjectServiceImpl extends ProjectServiceImplBase {

    private final CreateProjectService createProjectServiceImpl;

    @Override
    public void createProject(CreateProjectRequest request, StreamObserver<CreateProjectResponse> responseObserver) {
        createProjectServiceImpl.createProject(request, responseObserver);
    }

    [...]

}
````

3 ) Would look roughly like this:

````java
@GrpcService
public class MainProjectServiceImpl implements BindableService {

    public final ServerServiceDefinition bindService() {
        return ServerServiceDefinition.builder(ProjectServiceGrpc.getServiceDescriptor())
            .addMethod(createProjectServiceImpl.bindService().getMethod("createProject"))
            [...]
            .build();
    }

}
````

I would recommend variant 2 as it is probably the least confusing one.

-------------------------------

Does this solve your issue?

**Comment by ST-DDT** [comment]
Waiting for feedback
